docs/adr/0008-freeing-objects-and-resources.md

```diff
@@ -0,0 +1,105 @@
+# Freeing objects and resources
+
+* Status: Propsed
+* Deciders: ?
+* Date: 2023-10-10
+
+Prevous discussion: [PR 1394](https://github.com/mozilla/uniffi-rs/issues/1394), [Issue 8](https://github.com/mozilla/uniffi-rs/issues/8)
+
+## Context and Problem Statement
+
+UniFFI must support objects that hold resources that should be cleaned up promptly like database connections, file handles, network sockets, large memory allocations, etc.
```

Comment:
Thanks heaps for driving this forward Ben - it's a discussion we keep having that never seems to make progress!

Initially I'm just putting down some thoughts on the problem statement, as I believe tightening this up might help the rest.

#1394 says we could add "finalizer support to make a 'best effort' attempt at freeing what it can while leaving the auto-closeable pattern in place" - I'm not aware of us exploring that, nor how this ADR might change if that turned out to be successful.

This section starts off talking about all bindings in general, but then fairly quickly pivots into just Kotlin. While Kotlin is important, I think it's still worth a bit more of an explanation about how the other "important" bindings work. For example, it seems that Python also shares the same general problem - a context-manager is available, but this doesn't really work for long lived references. #8 implies Swift does not have this problem. C# and Go seem worth mentioning too as I suspect they might end up being very popular UniFFI bindings.

Upon re-reading this ADR, I think your "Option 1" can be restated as "Do nothing. It is up to the designer of the API exposed via UniFFI to take care of this. An option might be to add explicit "close()" methods, but that's all on you, we don't care". If that's accurate, maybe update the wording to be more explicit/brutal? We've also been talking about "transitive requirements (eg, enum holding a dictionary which holds a resource - there's nothing generic you can do on the enum to clean this tree up) - if the intention is also to offer no help there and leave that to the API designer it might be worth calling that out.

I also don't really understand the callback example - or more specifically, how the callback example is any different from the general problem. If you consider a pure-python program which takes some arbitrary object, there's also no generic way for the holder of that object to release the resources held by that arbitrary object - but that's not generally considered a problem - the solutions are either (a) know the type, so, eg, if it's a file you know you can .close() it, or (b) don't do that. IOW, the responsibility for this on the designer of the API so that either (a) the API has an explicit close-type method, or (b) don't require the Rust code to keep a long-lived reference to an object holding resources you care about?

tl;dr - I think I'm also in favour of Option 1 if I understand it correctly - and if I do, I'd suggest:
* Making that option even more explicit and "brutal" - ie, explicitly call out "this is your problem, UniFFI doesn't help"
* Explicitly call out the transitive case (ie, if your enums hold objects which manage resources, you are probably going to have a bad time, but UniFFI isn't going to try and stop you)
* Either remove most of the callback language (ie, replace it with something like "this is just a slightly different view of the exact same problem") or clarify how callbacks differ.
* Make the problem statement slightly less Kotlin specific, or explain how, eg, Python is different.

And probably most importantly, work out which Kotlin people to put this in-front of. Most UniFFI core contributors know enough Kotlin to be dangerous, but probably not quite enough to know how viable this will be in practice.

Reply:
Thanks, I think these are great suggestions.  I "brutalized" option 1 a bit, I could have gone farther, but I think it makes the point pretty well now.  I added some text about the transitive case and also other languages.  I agree it shouldn't be Kotlin specific.

I moved callbacks to their own section and tried to elaborate a bit more.  On the one hand, I think that it's all part of the same problem, but I also think they're important because they provide extra complications for any automatically generated solution.

Next step is to find some more Kotlin people...